lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp

```diff
@@ -0,0 +1,173 @@
+  }
+}
+
+struct LowerPackOnInputsWithoutTranspose
+    : public OpRewritePattern<linalg::GenericOp> {
+  // Is only called with single-user packOp operands, so callback can always
+  // find the (use by the) linalg.generic that is the target of the pattern.
+  using ControlFn = std::function<bool(tensor::PackOp)>;
+  ControlFn controlFn;
+
+  LowerPackOnInputsWithoutTranspose(MLIRContext *context,
+                                    ControlFn controlFn = nullptr,
+                                    PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit), controlFn(std::move(controlFn)) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    bool modifiedAnOperand = false;
+    for (auto operandIdx : {0, 1}) {
+      auto packOp = dyn_cast_if_present<tensor::PackOp>(
+          genericOp->getOperand(operandIdx).getDefiningOp());
+
+      if (!packOp || !packOp->hasOneUse() || (controlFn && !controlFn(packOp)))
+        continue;
+
+      lowerPackWithoutTranspose(packOp, genericOp, operandIdx, rewriter);
+
+      modifiedAnOperand = true;
+    }
+
+    return modifiedAnOperand ? success() : failure();
+  }
+};
+
+struct LowerPackUnpackOnOutputWithoutTranspose
+    : public OpRewritePattern<linalg::GenericOp> {
+  // Is only called with single-user packOp operands, so callback can always
+  // find the (use by the) linalg.generic that is the target of the pattern.
+  using ControlFn = std::function<bool(tensor::PackOp, tensor::UnPackOp)>;
+  ControlFn controlFn;
+
+  LowerPackUnpackOnOutputWithoutTranspose(MLIRContext *context,
+                                          ControlFn controlFn = nullptr,
+                                          PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit), controlFn(std::move(controlFn)) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!genericOp->hasOneUse())
+      return failure();
+
+    auto packOp = dyn_cast_if_present<tensor::PackOp>(
+        genericOp->getOperand(2).getDefiningOp());
+    auto unpackOp = llvm::dyn_cast<tensor::UnPackOp>(
+        *(genericOp->getResult(0).getUsers().begin()));
+
+    if (!packOp || !packOp->hasOneUse() || !unpackOp ||
+        (controlFn && !controlFn(packOp, unpackOp)))
+      return failure();
+
+    auto unpackDest = unpackOp.getDest();
+    bool destHasStaticShape = unpackDest.getType().hasStaticShape();
+
+    lowerPackWithoutTranspose(packOp, genericOp, /*operandIdx=*/2, rewriter);
+    auto res = linalg::lowerUnPack(rewriter, unpackOp);
+
+    // Set genericOp's result type to the adjusted type of the out parameter.
+    genericOp->getOpResult(0).setType(genericOp.getOperand(2).getType());
+
+    if (auto transposeOp = res->transposeOp) {
+      // Forget about the transpose introduced by lowerUnPack.
```

Comment:
This might not be safe to do if unpack has different permutations than pack.

Reply:
Or to put it more simply if you have sth like:
```
%packed = pack(<128x64>)
generic outs(%packed)
unpack into <64x128>
```

Reply:
Have made it such that the pack on the out param and the unpack on the corresponding result must specify the same permutation. Else the pattern does not apply.